src/closed.jl

```diff
@@ -0,0 +1,45 @@
+"""
+Defines the ClosedSet type.
+
+A ClosedSet(left, right) is an interval set where both bounds are rightd i.e. the range is [left, right]
+"""
+immutable ClosedSet{T}
+	left::T
+	right::T
+end
+
+function ClosedSet(left, right)
+	(I_left, I_right) = promote(left, right)
+	ClosedSet(I_left, I_right)
+end
+
+show(io::IO, I::ClosedSet) = print(io, "[", I.left, ",", I.right, "]")
+
+length(I::ClosedSet) = I.right - I.left + 1
+
+in(v, I::ClosedSet) = I.left <= v <= I.right
+
+isempty(A::ClosedSet) = A.left > A.right
+
+isequal(A::ClosedSet, B::ClosedSet) = (A.left == B.left && A.right == B.right) || isempty(A) && isempty(B)
+
+==(A::ClosedSet, B::ClosedSet) = isequal(A, B)
+!=(A::ClosedSet, B::ClosedSet) = !isequal(A, B)
+
+function intersect(A::ClosedSet, B::ClosedSet)
+	left = max(A.left, B.left)
+	right = min(A.right, B.right)
+	left, right = promote(left, right)
+	ClosedSet(left, right)
+end
+
+function union(A::ClosedSet, B::ClosedSet)
+	left = min(A.left, B.left)
+	right = max(A.right, B.right)
+	left, right = promote(left, right)
+	ClosedSet(left, right)
+end
```

Comment:
`intersect` doesn't have to, just add some tests for intersection of empty sets and I bet you'll see it already works as-is.

More importantly, `union` needs to make sure that the intervals intersect; I think this implementation will "fill in" gaps between two disjoint intervals. It should throw an error (`InexactError`? `ArgumentError`?) when the two inputs don't intersect, because the union can't be expressed as a single closed set.